README.md
# node-red-contrib-iiot-rpi-pca9685

[![platform](https://img.shields.io/badge/platform-Node--RED-red)](https://nodered.org)
[![platform](https://img.shields.io/badge/platform-Raspberry--Pi-ff69b4)](https://www.raspberrypi.com/)
![NPM version](https://badge.fury.io/js/node-red-contrib-iiot-rpi-pca9685.svg)
![NPM](https://img.shields.io/npm/l/node-red-contrib-iiot-rpi-pca9685)
[![Donate](https://img.shields.io/badge/Donate-PayPal-yellow.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=ZDRCZBQFWV3A6)

A Node-Red node for pca9685 16-Channel pwm/servo driver<br>

![image info](images/pca9685.png)

Example  **pca9685** hat from [WAVESHARE](https://www.waveshare.com/servo-driver-hat.htm)

![image info](images/pca9685module.png)

Example **pca9685** module

## Nodes
- **pca9685-pwm**: Generate pwm.
- **pca9685-svr**: Controls analog or digital servo.

## Installation
Install with Node-Red Palette Manager or npm command:
```
cd ~/.node-red
npm install node-red-contrib-iiot-rpi-pca9685
```

## Usage
- This node works on Raspberry Pi with 32bit or 64bit OS.
- PWM frequency 24Hz to 1526Hz.
- Servo timing can be set.
- Servo frequency 50Hz or 290Hz.
- Consult servo manual for timing.
- Enable I2C with raspi-config.

[CHANGELOG](CHANGELOG.md)<br>

CHANGELOG.md
1.0.21
- Initial Release


lib/syslib.js
/**
 * Copyright 2022 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use node file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

const fs = require('fs');

module.exports.outError = function (node, errShort, errLong) {
	if (node.save_txt === errShort)
		return true;

	node.save_txt = errShort;
	node.status({ fill: "red", shape: "ring", text: errShort });
	node.error(errLong);

	return true;
}

module.exports.outText = function (node, txt) {
	if (node.save_txt === txt)
		return;

	node.save_txt = txt;
	node.status({ text: txt });
}

function getRpiTarget()
{
	if (process.platform !== 'linux')
		return undefined;
	
	var cpuinfo = fs.readFileSync("/proc/cpuinfo").toString();

	if (cpuinfo.indexOf(": BCM") === -1)
		return undefined;
	
	if (process.arch === "arm64")
		return "_s64.node";
	else if (process.arch === "arm")
		return "_s32.node";
	
	return undefined;
}

module.exports.LoadModule = function (module) {
	try {
		var modName = getRpiTarget();
		if (modName !== undefined)
			return require("./" + module + modName);
	}
	catch (e) { console.log(e); }
	return undefined;
}


pca9685-srv.js
/**
 * Copyright 2022 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

module.exports = function(RED) {
	const syslib = require("./lib/syslib.js");
	const sysmodule = syslib.LoadModule("rpi_pca9685");

    RED.nodes.registerType("pca9685-srv", function(n) {
		var node = this;
		RED.nodes.createNode(node, n);

		node.devadr  = parseInt(n.devadr);
		node.channel = parseInt(n.channel);
		node.tlow  = n.tlow;
		node.thigh = n.thigh;
		node.pos = n.pos;
		node.digital = n.digital; 
		node.iserror = false;
		node.name = n.name || "servo @" + (0x40 + parseInt(node.devadr)).toString(16).toUpperCase() + "#" + node.channel;

		if (sysmodule === undefined)
			node.iserror = syslib.outError(node, "driver error", "driver not load, wrong os or not Raspi");
		else if (!sysmodule.initservo(node.devadr, node.digital))
			node.iserror = syslib.outError(node, "device used", "device used by other node or i2c not open");
		else if (!sysmodule.setservo(node.devadr, node.channel, node.pos, node.tlow, node.thigh))
			syslib.outError(node, "write error", "device not write, check i2c and device");
		else
			syslib.outText(node, node.pos);

		node.on("input", function (msg) {
			if (node.iserror)
				return;

			var val = Number(msg.payload);

			if (isNaN(val)) {
				syslib.outError(node, "not number", "msg.payload not number");
				return;
			}
			
			if (val < 0)
				val = 0;
			else if (val > 100)
				val = 100;
	
			if (!sysmodule.setservo(node.devadr, node.channel, val, node.tlow, node.thigh))
				syslib.outError(node, "write error", "device not write, check i2c and device");
			else
				syslib.outText(node, val);
		});

		node.on('close', function () {
			sysmodule.deinit(node.devadr);
		});
	});
}


pca9685-pwm.js
/**
 * Copyright 2022 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

module.exports = function(RED) {
	const syslib = require("./lib/syslib.js");
	const sysmodule = syslib.LoadModule("rpi_pca9685");

    RED.nodes.registerType("pca9685-pwm", function(n) {
		var node = this;
		RED.nodes.createNode(node, n);

		node.devadr = parseInt(n.devadr);
		node.channel = n.channel;
		node.freq = n.freq;
		node.duty = n.duty;
		node.opendrain = n.opendrain; 
		node.iserror = false;
		node.name = n.name || "pca9685 @" + (0x40 + parseInt(node.devadr)).toString(16).toUpperCase() + "#" + node.channel;

		if (sysmodule === undefined)
			node.iserror = syslib.outError(node, "driver error", "driver not load, wrong os or not Raspi");
		else if (!sysmodule.initpwm(node.devadr, node.freq, node.opendrain))
			node.iserror = syslib.outError(node, "device used", "device used by other node or i2c not open");
		else if (!sysmodule.setpwm(node.devadr, node.channel, node.duty))
			syslib.outError(node, "write error", "device not write, check i2c and device");
		else
			syslib.outText(node, node.duty + "% - " + sysmodule.getfreq(node.devadr) + "Hz");

		node.on("input", function (msg) {
			if (node.iserror)
				return;

			var val = Number(msg.payload);

			if (isNaN(val)) {
				syslib.outError(node, "not number", "msg.payload not number");
				return;
			}
			
			if (val < 0)
				val = 0;
			else if (val > 100)
				val = 100;
	
			if (!sysmodule.setpwm(node.devadr, node.channel, val))
				syslib.outError(node, "write error", "device not write, check i2c and device");
			else
				syslib.outText(node, val + "% - " + sysmodule.getfreq(node.devadr) + "Hz");

		});

		node.on('close', function () {
			sysmodule.deinit(node.devadr);
		});
	});
}
